File: update_dashboard.py
#!/usr/bin/env python3
"""
Script de Automação - Dashboard CRM
Processa dados de planilhas Excel e atualiza o data.json no GitHub automaticamente
"""

import pandas as pd
import json
import os
import subprocess
from datetime import datetime
from pathlib import Path

class DashboardUpdater:
    def __init__(self, bourbon_file, cx_file, output_file='data.json'):
        """
        Inicializa o atualizador de dashboard
        
        Args:
            bourbon_file (str): Caminho para a planilha Bourbon
            cx_file (str): Caminho para a planilha CX Curadoria
            output_file (str): Nome do arquivo JSON de saída
        """
        self.bourbon_file = bourbon_file
        self.cx_file = cx_file
        self.output_file = output_file
        self.dashboard_data = {}
        
    def process_media_data(self):
        """Processa dados de mídia (Google Ads e Meta Ads)"""
        print("📊 Processando dados de mídia...")
        
        # Ler Google Ads
        google_ads = pd.read_excel(self.bourbon_file, sheet_name="Base Bourbon - Google Ads")
        google_ads_clean = google_ads.dropna(subset=['Year & month', 'Cost', 'Conversions']).copy()
        google_ads_clean['Channel'] = 'Google Ads'
        google_ads_clean['Cost'] = pd.to_numeric(google_ads_clean['Cost'], errors='coerce')
        google_ads_clean['Conversions'] = pd.to_numeric(google_ads_clean['Conversions'], errors='coerce')
        google_ads_clean['Receita GA4'] = pd.to_numeric(google_ads_clean['Receita GA4'], errors='coerce')
        
        # Ler Meta Ads
        meta_ads = pd.read_excel(self.bourbon_file, sheet_name="Base Bourbon - Meta Ads")
        meta_ads_clean = meta_ads.dropna(subset=['Year & month', 'Cost']).copy()
        meta_ads_clean['Channel'] = 'Meta Ads'
        meta_ads_clean['Cost'] = pd.to_numeric(meta_ads_clean['Cost'], errors='coerce')
        meta_ads_clean['Website purchases'] = pd.to_numeric(meta_ads_clean['Website purchases'], errors='coerce')
        meta_ads_clean['Receita GA4'] = pd.to_numeric(meta_ads_clean['Receita GA4'], errors='coerce')
        meta_ads_clean = meta_ads_clean[meta_ads_clean['Cost'] > 0]
        
        # Combinar dados
        media_combined = pd.concat([
            google_ads_clean[['Year & month', 'Channel', 'Cost', 'Conversions', 'Receita GA4']].rename(
                columns={'Conversions': 'Conversions_count'}),
            meta_ads_clean[['Year & month', 'Channel', 'Cost', 'Website purchases', 'Receita GA4']].rename(
                columns={'Website purchases': 'Conversions_count'})
        ], ignore_index=True)
        
        # Agrupar por período e canal
        media_by_period_channel = media_combined.groupby(['Year & month', 'Channel']).agg({
            'Cost': 'sum',
            'Conversions_count': 'sum',
            'Receita GA4': 'sum'
        }).reset_index()
        
        media_by_period_channel['CAC'] = media_by_period_channel['Cost'] / media_by_period_channel['Conversions_count'].replace(0, 1)
        media_by_period_channel['ROAS'] = media_by_period_channel['Receita GA4'] / media_by_period_channel['Cost'].replace(0, 1)
        
        print(f"✓ {len(media_by_period_channel)} registros de mídia processados")
        return media_by_period_channel
    
    def process_cx_data(self):
        """Processa dados de CX (NPS, CSAT, Nota IA)"""
        print("😊 Processando dados de CX...")
        
        cx_curadoria = pd.read_excel(self.cx_file, sheet_name="CX_Curadoria")
        cx_curadoria['Data'] = pd.to_datetime(cx_curadoria['Data'])
        cx_curadoria['Ano_Mes'] = cx_curadoria['Data'].dt.strftime('%Y|%m')
        cx_curadoria['Nota IA'] = pd.to_numeric(cx_curadoria['Nota IA'], errors='coerce')
        
        # Mapear NPS e CSAT
        nps_map = {'Promotor': 10, 'Neutro': 5, 'Detrator': 0}
        csat_map = {'Satisfeito': 10, 'Neutro': 5, 'Insatisfeito': 0}
        
        cx_curadoria['NPS_value'] = cx_curadoria['NPS P2'].map(nps_map)
        cx_curadoria['CSAT_value'] = cx_curadoria['CSAT P2'].map(csat_map)
        
        # Agrupar por período e canal
        cx_by_period_channel = cx_curadoria.groupby(['Ano_Mes', 'Canal']).agg({
            'NPS_value': 'mean',
            'CSAT_value': 'mean',
            'Nota IA': 'mean'
        }).reset_index()
        
        cx_by_period_channel.columns = ['Ano_Mes', 'Canal', 'NPS', 'CSAT', 'Nota_IA']
        
        print(f"✓ {len(cx_by_period_channel)} registros de CX processados")
        return cx_by_period_channel
    
    def create_dashboard_json(self, media_df, cx_df):
        """Cria estrutura JSON para o dashboard"""
        print("📦 Criando estrutura JSON...")
        
        self.dashboard_data = {
            'media_periods': sorted(media_df['Year & month'].unique().tolist()),
            'cx_periods': sorted(cx_df['Ano_Mes'].unique().tolist()),
            'all_channels': sorted(list(set(
                media_df['Channel'].unique().tolist() + 
                cx_df['Canal'].unique().tolist()
            ))),
            'media_by_period_channel': [],
            'cx_by_period_channel': [],
            'last_updated': datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        }
        
        # Adicionar dados de mídia
        for _, row in media_df.iterrows():
            self.dashboard_data['media_by_period_channel'].append({
                'period': str(row['Year & month']),
                'channel': row['Channel'],
                'cost': round(float(row['Cost']), 2),
                'conversions': round(float(row['Conversions_count']), 2),
                'revenue': round(float(row['Receita GA4']), 2),
                'cac': round(float(row['CAC']), 2),
                'roas': round(float(row['ROAS']), 2)
            })
        
        # Adicionar dados de CX
        for _, row in cx_df.iterrows():
            self.dashboard_data['cx_by_period_channel'].append({
                'period': str(row['Ano_Mes']),
                'channel': row['Canal'],
                'nps': round(float(row['NPS']), 2),
                'csat': round(float(row['CSAT']), 2),
                'nota_ia': round(float(row['Nota_IA']), 2)
            })
        
        print(f"✓ JSON criado com {len(self.dashboard_data['media_by_period_channel'])} registros de mídia e {len(self.dashboard_data['cx_by_period_channel'])} registros de CX")
    
    def save_json(self):
        """Salva JSON no arquivo"""
        print(f"💾 Salvando {self.output_file}...")
        
        with open(self.output_file, 'w', encoding='utf-8') as f:
            json.dump(self.dashboard_data, f, indent=2, ensure_ascii=False)
        
        print(f"✓ Arquivo salvo: {self.output_file}")
    
    def git_commit_push(self, commit_message=None):
        """Faz commit e push para o GitHub"""
        print("🚀 Enviando para o GitHub...")
        
        if commit_message is None:
            commit_message = f"Atualização automática dos dados - {datetime.now().strftime('%Y-%m-%d %H:%M')}"
        
        try:
            # Add
            subprocess.run(['git', 'add', self.output_file], check=True)
            
            # Commit
            subprocess.run(['git', 'commit', '-m', commit_message], check=True)
            
            # Push
            subprocess.run(['git', 'push', 'origin', 'main'], check=True)
            
            print("✓ Dados enviados para o GitHub com sucesso!")
            print("⏳ Aguarde 2-5 minutos para o GitHub Pages atualizar")
            
        except subprocess.CalledProcessError as e:
            print(f"❌ Erro ao enviar para o GitHub: {e}")
            print("Certifique-se de que você está no diretório do repositório e tem permissões de push")
    
    def run(self, auto_push=True):
        """Executa o processo completo de atualização"""
        print("=" * 60)
        print("🔄 INICIANDO ATUALIZAÇÃO DO DASHBOARD")
        print("=" * 60)
        
        try:
            # Processar dados
            media_df = self.process_media_data()
            cx_df = self.process_cx_data()
            
            # Criar JSON
            self.create_dashboard_json(media_df, cx_df)
            
            # Salvar
            self.save_json()
            
            # Push para GitHub
            if auto_push:
                self.git_commit_push()
            
            print("=" * 60)
            print("✅ ATUALIZAÇÃO CONCLUÍDA COM SUCESSO!")
            print("=" * 60)
            
            return True
            
        except Exception as e:
            print(f"❌ Erro durante a atualização: {e}")
            import traceback
            traceback.print_exc()
            return False


def main():
    """Função principal"""
    import argparse
    
    parser = argparse.ArgumentParser(description='Atualiza o dashboard CRM automaticamente')
    parser.add_argument('--bourbon', required=True, help='Caminho para a planilha Bourbon')
    parser.add_argument('--cx', required=True, help='Caminho para a planilha CX Curadoria')
    parser.add_argument('--output', default='data.json', help='Nome do arquivo JSON de saída')
    parser.add_argument('--no-push', action='store_true', help='Não fazer push para o GitHub')
    
    args = parser.parse_args()
    
    # Verificar se os arquivos existem
    if not os.path.exists(args.bourbon):
        print(f"❌ Arquivo não encontrado: {args.bourbon}")
        return
    
    if not os.path.exists(args.cx):
        print(f"❌ Arquivo não encontrado: {args.cx}")
        return
    
    # Executar atualização
    updater = DashboardUpdater(args.bourbon, args.cx, args.output)
    success = updater.run(auto_push=not args.no_push)
    
    if success:
        print("\n📊 Dashboard disponível em: https://marcoscoelhot4c.github.io/dashboard-crm/")
    else:
        print("\n❌ Atualização falhou. Verifique os erros acima.")


if __name__ == '__main__':
    main()

File: update_dashboard.js
#!/usr/bin/env node
/**
 * Script de Automação - Dashboard CRM (Node.js)
 * Processa dados de planilhas Excel e atualiza o data.json no GitHub automaticamente
 */

const XLSX = require('xlsx');
const fs = require('fs');
const { execSync } = require('child_process');
const path = require('path');

class DashboardUpdater {
    constructor(bourbonFile, cxFile, outputFile = 'data.json') {
        this.bourbonFile = bourbonFile;
        this.cxFile = cxFile;
        this.outputFile = outputFile;
        this.dashboardData = {};
    }

    /**
     * Processa dados de mídia (Google Ads e Meta Ads)
     */
    processMediaData() {
        console.log('📊 Processando dados de mídia...');

        const workbook = XLSX.readFile(this.bourbonFile);

        // Ler Google Ads
        const googleAdsSheet = workbook.Sheets['Base Bourbon - Google Ads'];
        const googleAds = XLSX.utils.sheet_to_json(googleAdsSheet);
        
        const googleAdsClean = googleAds
            .filter(row => row['Year & month'] && row['Cost'] && row['Conversions'])
            .map(row => ({
                period: row['Year & month'],
                channel: 'Google Ads',
                cost: parseFloat(row['Cost']) || 0,
                conversions: parseFloat(row['Conversions']) || 0,
                revenue: parseFloat(row['Receita GA4']) || 0
            }));

        // Ler Meta Ads
        const metaAdsSheet = workbook.Sheets['Base Bourbon - Meta Ads'];
        const metaAds = XLSX.utils.sheet_to_json(metaAdsSheet);
        
        const metaAdsClean = metaAds
            .filter(row => row['Year & month'] && row['Cost'] && parseFloat(row['Cost']) > 0)
            .map(row => ({
                period: row['Year & month'],
                channel: 'Meta Ads',
                cost: parseFloat(row['Cost']) || 0,
                conversions: parseFloat(row['Website purchases']) || 0,
                revenue: parseFloat(row['Receita GA4']) || 0
            }));

        // Combinar dados
        const mediaCombined = [...googleAdsClean, ...metaAdsClean];

        // Agrupar por período e canal
        const mediaGrouped = {};
        mediaCombined.forEach(item => {
            const key = `${item.period}|${item.channel}`;
            if (!mediaGrouped[key]) {
                mediaGrouped[key] = {
                    period: item.period,
                    channel: item.channel,
                    cost: 0,
                    conversions: 0,
                    revenue: 0
                };
            }
            mediaGrouped[key].cost += item.cost;
            mediaGrouped[key].conversions += item.conversions;
            mediaGrouped[key].revenue += item.revenue;
        });

        // Calcular CAC e ROAS
        const mediaByPeriodChannel = Object.values(mediaGrouped).map(item => ({
            period: item.period,
            channel: item.channel,
            cost: Math.round(item.cost * 100) / 100,
            conversions: Math.round(item.conversions * 100) / 100,
            revenue: Math.round(item.revenue * 100) / 100,
            cac: Math.round((item.cost / (item.conversions || 1)) * 100) / 100,
            roas: Math.round((item.revenue / (item.cost || 1)) * 100) / 100
        }));

        console.log(`✓ ${mediaByPeriodChannel.length} registros de mídia processados`);
        return mediaByPeriodChannel;
    }

    /**
     * Processa dados de CX (NPS, CSAT, Nota IA)
     */
    processCXData() {
        console.log('😊 Processando dados de CX...');

        const workbook = XLSX.readFile(this.cxFile);
        const cxSheet = workbook.Sheets['CX_Curadoria'];
        const cxData = XLSX.utils.sheet_to_json(cxSheet);

        // Mapear NPS e CSAT
        const npsMap = { 'Promotor': 10, 'Neutro': 5, 'Detrator': 0 };
        const csatMap = { 'Satisfeito': 10, 'Neutro': 5, 'Insatisfeito': 0 };

        // Processar dados
        const cxProcessed = cxData
            .filter(row => row['Data'] && row['Canal'])
            .map(row => {
                const date = new Date(row['Data']);
                const year = date.getFullYear();
                const month = String(date.getMonth() + 1).padStart(2, '0');
                const period = `${year}|${month}`;

                return {
                    period: period,
                    channel: row['Canal'],
                    nps: npsMap[row['NPS P2']] || 0,
                    csat: csatMap[row['CSAT P2']] || 0,
                    nota_ia: parseFloat(row['Nota IA']) || 0
                };
            });

        // Agrupar por período e canal
        const cxGrouped = {};
        cxProcessed.forEach(item => {
            const key = `${item.period}|${item.channel}`;
            if (!cxGrouped[key]) {
                cxGrouped[key] = {
                    period: item.period,
                    channel: item.channel,
                    nps: [],
                    csat: [],
                    nota_ia: []
                };
            }
            cxGrouped[key].nps.push(item.nps);
            cxGrouped[key].csat.push(item.csat);
            cxGrouped[key].nota_ia.push(item.nota_ia);
        });

        // Calcular médias
        const cxByPeriodChannel = Object.values(cxGrouped).map(item => ({
            period: item.period,
            channel: item.channel,
            nps: Math.round((item.nps.reduce((a, b) => a + b, 0) / item.nps.length) * 100) / 100,
            csat: Math.round((item.csat.reduce((a, b) => a + b, 0) / item.csat.length) * 100) / 100,
            nota_ia: Math.round((item.nota_ia.reduce((a, b) => a + b, 0) / item.nota_ia.length) * 100) / 100
        }));

        console.log(`✓ ${cxByPeriodChannel.length} registros de CX processados`);
        return cxByPeriodChannel;
    }

    /**
     * Cria estrutura JSON para o dashboard
     */
    createDashboardJSON(mediaData, cxData) {
        console.log('📦 Criando estrutura JSON...');

        const mediaPeriods = [...new Set(mediaData.map(d => d.period))].sort();
        const cxPeriods = [...new Set(cxData.map(d => d.period))].sort();
        const allChannels = [...new Set([
            ...mediaData.map(d => d.channel),
            ...cxData.map(d => d.channel)
        ])].sort();

        this.dashboardData = {
            media_periods: mediaPeriods,
            cx_periods: cxPeriods,
            all_channels: allChannels,
            media_by_period_channel: mediaData,
            cx_by_period_channel: cxData,
            last_updated: new Date().toISOString()
        };

        console.log(`✓ JSON criado com ${mediaData.length} registros de mídia e ${cxData.length} registros de CX`);
    }

    /**
     * Salva JSON no arquivo
     */
    saveJSON() {
        console.log(`💾 Salvando ${this.outputFile}...`);

        fs.writeFileSync(
            this.outputFile,
            JSON.stringify(this.dashboardData, null, 2),
            'utf-8'
        );

        console.log(`✓ Arquivo salvo: ${this.outputFile}`);
    }

    /**
     * Faz commit e push para o GitHub
     */
    gitCommitPush(commitMessage = null) {
        console.log('🚀 Enviando para o GitHub...');

        if (!commitMessage) {
            const now = new Date();
            commitMessage = `Atualização automática dos dados - ${now.toISOString().split('T')[0]} ${now.toTimeString().split(' ')[0]}`;
        }

        try {
            // Add
            execSync(`git add ${this.outputFile}`, { stdio: 'inherit' });

            // Commit
            execSync(`git commit -m "${commitMessage}"`, { stdio: 'inherit' });

            // Push
            execSync('git push origin main', { stdio: 'inherit' });

            console.log('✓ Dados enviados para o GitHub com sucesso!');
            console.log('⏳ Aguarde 2-5 minutos para o GitHub Pages atualizar');

        } catch (error) {
            console.error('❌ Erro ao enviar para o GitHub:', error.message);
            console.error('Certifique-se de que você está no diretório do repositório e tem permissões de push');
        }
    }

    /**
     * Executa o processo completo de atualização
     */
    run(autoPush = true) {
        console.log('='.repeat(60));
        console.log('🔄 INICIANDO ATUALIZAÇÃO DO DASHBOARD');
        console.log('='.repeat(60));

        try {
            // Processar dados
            const mediaData = this.processMediaData();
            const cxData = this.processCXData();

            // Criar JSON
            this.createDashboardJSON(mediaData, cxData);

            // Salvar
            this.saveJSON();

            // Push para GitHub
            if (autoPush) {
                this.gitCommitPush();
            }

            console.log('='.repeat(60));
            console.log('✅ ATUALIZAÇÃO CONCLUÍDA COM SUCESSO!');
            console.log('='.repeat(60));

            return true;

        } catch (error) {
            console.error('❌ Erro durante a atualização:', error);
            console.error(error.stack);
            return false;
        }
    }
}

// Função principal
function main() {
    const args = process.argv.slice(2);

    if (args.includes('--help') || args.includes('-h')) {
        console.log(`
Uso: node update_dashboard.js --bourbon <arquivo> --cx <arquivo> [opções]

Opções:
  --bourbon <arquivo>    Caminho para a planilha Bourbon (obrigatório)
  --cx <arquivo>         Caminho para a planilha CX Curadoria (obrigatório)
  --output <arquivo>     Nome do arquivo JSON de saída (padrão: data.json)
  --no-push              Não fazer push para o GitHub
  --help, -h             Mostra esta mensagem de ajuda

Exemplo:
  node update_dashboard.js --bourbon dados/bourbon.xlsx --cx dados/cx.xlsx
        `);
        return;
    }

    const bourbonIndex = args.indexOf('--bourbon');
    const cxIndex = args.indexOf('--cx');
    const outputIndex = args.indexOf('--output');
    const noPush = args.includes('--no-push');

    if (bourbonIndex === -1 || cxIndex === -1) {
        console.error('❌ Erro: --bourbon e --cx são obrigatórios');
        console.error('Use --help para ver as opções disponíveis');
        process.exit(1);
    }

    const bourbonFile = args[bourbonIndex + 1];
    const cxFile = args[cxIndex + 1];
    const outputFile = outputIndex !== -1 ? args[outputIndex + 1] : 'data.json';

    // Verificar se os arquivos existem
    if (!fs.existsSync(bourbonFile)) {
        console.error(`❌ Arquivo não encontrado: ${bourbonFile}`);
        process.exit(1);
    }

    if (!fs.existsSync(cxFile)) {
        console.error(`❌ Arquivo não encontrado: ${cxFile}`);
        process.exit(1);
    }

    // Executar atualização
    const updater = new DashboardUpdater(bourbonFile, cxFile, outputFile);
    const success = updater.run(!noPush);

    if (success) {
        console.log('\n📊 Dashboard disponível em: https://marcoscoelhot4c.github.io/dashboard-crm/');
    } else {
        console.log('\n❌ Atualização falhou. Verifique os erros acima.');
        process.exit(1);
    }
}

// Executar se for chamado diretamente
if (require.main === module) {
    main();
}

module.exports = DashboardUpdater;
